Mathlib/Tactic/Basic.lean

```diff
@@ -54,3 +54,69 @@ set_option hygiene false in
 macro_rules
   | `(tactic| transitivity) => `(tactic| apply Nat.lt_trans)
   | `(tactic| transitivity $e) => `(tactic| apply Nat.lt_trans (m := $e))
+
+syntax (name := introv) "introv " (colGt ident)* : tactic
+@[tactic introv] partial def evalIntrov : Tactic := fun stx => do
+  match stx with
+  | `(tactic| introv)                    => introsDep
+  | `(tactic| introv $h:ident $hs:ident*) => evalTactic (← `(tactic| introv; intro $h:ident; introv $hs:ident*))
+  | _ => throwUnsupportedSyntax
+where
+  introsDep : TacticM Unit := do
+    let t ← getMainTarget
+    match t with
+    | Expr.forallE _ _ e _ =>
+      if e.hasLooseBVars then
+        intro1PStep
+        introsDep
+    | _ => ()
+  intro1PStep : TacticM Unit :=
+    liftMetaTactic fun mvarId => do
+      let (_, mvarId) ← Meta.intro1P mvarId
+      pure [mvarId]
+
+syntax (name := byContra) "byContra " (colGt ident)? : tactic
+macro_rules
+  | `(tactic| byContra) => `(tactic| (apply Decidable.byContradiction; intro))
+  | `(tactic| byContra $e) => `(tactic| (apply Decidable.byContradiction; intro $e))
+macro_rules
+  | `(tactic| byContra) => `(tactic| (apply Classical.byContradiction; intro))
+  | `(tactic| byContra $e) => `(tactic| (apply Classical.byContradiction; intro $e))
+
+syntax (name := guardExprEq) "guardExprEq " term " := " term : tactic
+@[tactic guardExprEq] def evalGuardExprEq : Lean.Elab.Tactic.Tactic := fun stx => do
+  let r ← elabTerm stx[1] none
+  let p ← elabTerm stx[3] none
+  if not (r == p) then throwError m!"failed"
```

Comment:
You can usually use the elab command to combine the syntax definition and elaborator:

```suggestion
elab "guardExprEq " r:term " := " p:term : tactic => do
  let r ← elabTerm r none
  let p ← elabTerm p none
  if not (r == p) then throwError "failed: {r} != {p}"
```

The `m!` isn't necessary after throwError either, since it's a macro which automatically adds the `m!` if the argument is a string.